crates/ra_assists/src/merge_match_arms.rs

```diff
@@ -0,0 +1,196 @@
+use hir::db::HirDatabase;
+use ra_syntax::{
+    ast::{MatchArm, AstNode,Pat,Whitespace,AstToken},
+    SyntaxElement,
+    Direction,
+    SyntaxKind::COMMA,
+    TextUnit,
+    algo::non_trivia_sibling
+};
+
+use crate::{AssistCtx, Assist, AssistId};
+
+pub(crate) fn merge_match_arms(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let match_arm = ctx.node_at_offset::<MatchArm>().unwrap();
+
+    let next_arm = MatchArm::cast(match_arm.syntax().next_sibling()?)?;
+
+    let match_arm_expr = match_arm.expr()?;
+
+    let next_arm_expr = next_arm.expr()?;
+
+    if match_arm_expr.syntax().text().len() != next_arm_expr.syntax().text().len()
+        && match_arm_expr.syntax().text() != next_arm_expr.syntax().text()
+    {
+        return None;
+    }
+
+    let buf = build_buffer(match_arm.pats());
+
+    let comma = non_trivia_sibling(match_arm.syntax().into(), Direction::Next);
+
+    let white_space_offset = match_arm.syntax().prev_sibling_or_token();
+
+    ctx.add_action(AssistId("merge_match_arms"), "merge match arms", |edit| {
+        edit.target(match_arm.syntax().range());
+
+        let offseting_amount = match white_space_offset {
+            Some(SyntaxElement::Token(tok)) => {
+                if let Some(_) = Whitespace::cast(tok) {
+                    let ele = white_space_offset.unwrap().range();
+                    ele.len()
+                } else {
+                    TextUnit::from(0)
+                }
+            }
+            _ => TextUnit::from(0),
+        };
+
+        if let Some(comma) = comma {
+            if comma.kind() == COMMA {
+                edit.delete(comma.range());
+            }
+        }
+
+        edit.replace_node_and_indent(match_arm.syntax(), buf);
+
+        if let Some(comma) = non_trivia_sibling(match_arm.syntax().into(), Direction::Next) {
+            if comma.kind() == COMMA {
+                edit.set_cursor(
+                    match_arm.syntax().range().end()
+                        + offseting_amount
+                        + comma.range().len()
+                        + TextUnit::from(2),
+                );
+            } else {
+                edit.set_cursor(
+                    match_arm.syntax().range().end() + offseting_amount + TextUnit::from(3),
+                )
+            }
+        } else {
+            println!("{} vs {}", match_arm.syntax().range().end(), next_arm.syntax().range().end());
+            edit.set_cursor(match_arm.syntax().range().end() + offseting_amount)
+        }
+    });
+
+    ctx.build()
+}
+
+fn build_buffer<'a>(pats: impl Iterator<Item = &'a Pat>) -> String {
+    let mut buf = String::new();
+
+    pats.for_each(|pat| buf.push_str(&format!("{}", pat.syntax().text())));
+
+    buf.push_str(" |");
+
+    buf
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Let's add

* a test with a more complex expression than just `()`
* a test that checks that assits is not applicable if expressions are different
* a test with complex `|`-separated patterns